src/web_app/js/background.js

```diff
@@ -9,14 +9,151 @@
 /* More information about these options at jshint.com/docs/options */
 
 // Variables defined in and used from chrome.
-/* globals chrome */
+/* globals chrome, sendAsyncUrlRequest, Constants, parseJSON */
 
 'use strict';
-chrome.app.runtime.onLaunched.addListener(function() {
-  chrome.app.window.create('appwindow.html', {
-    'width': 800,
-    'height': 600,
-    'left': 0,
-    'top': 0
+(function() {
+  chrome.app.runtime.onLaunched.addListener(function() {
+    chrome.app.window.create('appwindow.html', {
+      'width': 800,
+      'height': 600,
+      'left': 0,
+      'top': 0
+    });
   });
-});
+
+  // Send event notification from background to window.
+  var sendWSEventMessageToWindow = function(port, wsEvent, data) {
+    var message = {
+        action: Constants.WS_ACTION,
+        wsAction: Constants.EVENT_ACTION,
+        wsEvent: wsEvent
+    };
+    if (data) {
+      message.data = data;
+    }
+    trace('B -> W: ' + JSON.stringify(message));
+    try {
+      port.postMessage(message);
+    } catch (ex) {
+      trace('Error sending message: ' + ex);
+    }
+  };
+
+  var handleWebSocketRequest = function(port, message) {
+    if (message.wsAction === Constants.WS_CREATE_ACTION) {
+      trace('RWS: creating web socket: ' + message.wssUrl);
+      var ws = new WebSocket(message.wssUrl);
+      port.wssPostUrl_ = message.wssPostUrl;
+      ws.onopen = function() {
+        trace('RWS: onopen');
+        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONOPEN);
+      };
+      ws.onerror = function() {
+        trace('RWS: onerror.');
+        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONERROR);
+      };
+      ws.onclose = function(event) {
+        trace('RWS: onclose - code: ' + event.code + ' reason: ' +
+            event.reason);
+        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONCLOSE, event);
+      };
+      ws.onmessage = function(event) {
+        trace('RWS: onmessage: ' + event.data);
+        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONMESSAGE, event);
+      };
+      port.webSocket_ = ws;
+    } else if (message.wsAction === Constants.WS_SEND_ACTION) {
+      trace('RWS: sending: ' + message.data);
+      if (port.webSocket_ && port.webSocket_.readyState === WebSocket.OPEN) {
+        try {
+          port.webSocket_.send(message.data);
+        } catch (ex) {
+          trace('RWS: exception sending: ' + ex);
```

Comment:
On Fri, Feb 20, 2015 at 2:51 PM, Chuck Hays <notifications@github.com>
wrote:

> In src/web_app/js/background.js
> <https://github.com/webrtc/apprtc/pull/82#discussion_r25110228>:
>
> > +        trace('RWS: onclose - code: ' + event.code + ' reason: ' +
> > +            event.reason);
> > +        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONCLOSE, event);
> > +      };
> > +      ws.onmessage = function(event) {
> > +        trace('RWS: onmessage: ' + event.data);
> > +        sendWSEventMessageToWindow(port, Constants.WS_EVENT_ONMESSAGE, event);
> > +      };
> > +      port.webSocket_ = ws;
> > +    } else if (message.wsAction === Constants.WS_SEND_ACTION) {
> > +      trace('RWS: sending: ' + message.data);
> > +      if (port.webSocket_ && port.webSocket_.readyState === WebSocket.OPEN) {
> > +        try {
> > +          port.webSocket_.send(message.data);
> > +        } catch (ex) {
> > +          trace('RWS: exception sending: ' + ex);
>
> I think that makes sense, right now we don't handle this type of exception
> in signalingchannel, but we should probably do the same thing in both
> places.
>
> I'm fine with doing it later if we didn't handle it before your change,
with a bug to track it.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/webrtc/apprtc/pull/82/files#r25110228>.
>



-- 
Jiayang Liu